RFM-2 Report

dennis-tra opened the pull request.

guillaumemichel commented:
Great report @dennis-tra ! Very good section about Measurement Methodology!

> For completeness, we present a selected number of graphs from the recent results (2022-05-05) in this report as well.

1. It would be insightful to have some more details about the origin of the data. What was the time period during which you collected data about the node classification? When and for how long did you run the churn measurements? 

> By diving into the data, many more questions arose that led to other RFMs like RFM5 or RFM6. Both of which can be tackeled with the the software that was developed in the scope of this RFM2.

2. It would be great to have a pointer to the descriptions of [RFM5](https://github.com/protocol/network-measurements/blob/master/RFMs.md#_fld2sr9emn3m) and [RFM6](https://github.com/protocol/network-measurements/blob/master/RFMs.md#_1xjatm7vfujn)

3. Add some references for `ioi` and `storm` nodes.

4. @yiannisbot I created a new branch [guillaumemichel-rfm-template](https://github.com/protocol/network-measurements/tree/guillaumemichel-rfm-template) with the repo structure we discussed. @dennis-tra Could you follow a similar structure? e.g write the report in `results/rfm2-uptime-and-churn.md`, create a folder `implementations/rfm2-uptime-and-churn/` containing the source code and/or scripts to reproduce the results and the plots used in the report (could be in a subfolder e.g `images/`, `resources` or similar)

    @yiannisbot @dennis-tra let me know if you have any comments on the structure, it can be adapted if needed :)

5. Minor typos: `tackeled` -> `tackled` [line 121](https://github.com/dennis-tra/network-measurements/blob/5bdbc00b19650c2ecc677ee30bf2d7c260e07391/reports/RFM2.md?plain=1#L121), `persistance` -> `persistence` [line 48](https://github.com/dennis-tra/network-measurements/blob/5bdbc00b19650c2ecc677ee30bf2d7c260e07391/reports/RFM2.md?plain=1#L48)

dennis-tra commented on reports/RFM2.md:
```diff
@@ -0,0 +1,131 @@
+
+A major performance indicator for an unstructured peer-to-peer network is participant churn, meaning how long do peers stay in the network after joining before going offline again. Network-wide parameters that cover the number of record replication and request concurrency rely on an accurate understanding of this dynamic. The average session length of a peer-to-peer network is a global measure of ...
+
+In the past, measurements on network churn have been conducted for networks like BitTorrent or Gnutella [^1] and only one team has briefly looked into churn of the Public IPFS Network [^2]. However, their data is dated by now as it was gathered before the release of DHT client vs. server node distinction in version `go-ipfs 0.5.0` and has not the same granularity as we are about to present. In th ...
```
Sebastians (and Leos) crawler does not distinguish between clients and servers (same as Nebula). They only see the DHT server nodes, so there's nothing to update :) What I wanted to point out is that the churn measurements were conducted before the 0.5.0 release. Hence, the topology and churn have likely changed and justifies a revisit of churn measurements 🤷‍♂️

dennis-tra commented:
Thanks @guillaumemichel and @yiannisbot for your feedback! I plan to add your requested changes today :)

> 4. @yiannisbot I created a new branch [guillaumemichel-rfm-template](https://github.com/protocol/network-measurements/tree/guillaumemichel-rfm-template) with the repo structure we discussed. @dennis-tra Could you follow a similar structure? e.g write the report in `results/rfm2-uptime-and-churn.md`, create a folder `implementations/rfm2-uptime-and-churn/` containing the source code and/or scripts to reproduce the results and the plots used in the report (could be in a subfolder e.g `images/`, `resources` or similar)

Sure, I'll restructure my PR :) I think though that a pointer to my repository or preferably a forked version from PL would be better suited for a source code/script reference - what do you think? Of course, I'm also happy to add the source code in this subdirectory :)

dennis-tra commented:
> Great report @dennis-tra ! Very good section about Measurement Methodology!
> 
> > For completeness, we present a selected number of graphs from the recent results (2022-05-05) in this report as well.
> 
> 1. It would be insightful to have some more details about the origin of the data. What was the time period during which you collected data about the node classification? When and for how long did you run the churn measurements?

Done 👍

> > By diving into the data, many more questions arose that led to other RFMs like RFM5 or RFM6. Both of which can be tackeled with the the software that was developed in the scope of this RFM2.
> 
> 2. It would be great to have a pointer to the descriptions of [RFM5](https://github.com/protocol/network-measurements/blob/master/RFMs.md#_fld2sr9emn3m) and [RFM6](https://github.com/protocol/network-measurements/blob/master/RFMs.md#_1xjatm7vfujn)

Done 👍

> 3. Add some references for `ioi` and `storm` nodes.

Suggestions welcome for an `ioi` reference - I haven't found any :( For storm, I found these ones:

- https://www.zdnet.com/article/ipstorm-botnet-expands-from-windows-to-android-mac-and-linux/
- https://www.anomali.com/blog/the-interplanetary-storm-new-malware-in-wild-using-interplanetary-file-systems-ipfs-p2p-network

what do you think? I actually haven't mentioned `storm` nodes in the report yet.

> 4. @yiannisbot I created a new branch [guillaumemichel-rfm-template](https://github.com/protocol/network-measurements/tree/guillaumemichel-rfm-template) with the repo structure we discussed. @dennis-tra Could you follow a similar structure? e.g write the report in `results/rfm2-uptime-and-churn.md`, create a folder `implementations/rfm2-uptime-and-churn/` containing the source code and/or scripts to reproduce the results and the plots used in the report (could be in a subfolder e.g `images/`, `resources` or similar)
>    @yiannisbot @dennis-tra let me know if you have any comments on the structure, it can be adapted if needed :)

I moved the report to `results/rfm2-uptime-and-churn.md`. I moved the graphs in a respective `results/rfm2-uptime-and-churn/` folder. Fine with you @guillaumemichel ?

> 5. Minor typos: `tackeled` -> `tackled` [line 121](https://github.com/dennis-tra/network-measurements/blob/5bdbc00b19650c2ecc677ee30bf2d7c260e07391/reports/RFM2.md?plain=1#L121), `persistance` -> `persistence` [line 48](https://github.com/dennis-tra/network-measurements/blob/5bdbc00b19650c2ecc677ee30bf2d7c260e07391/reports/RFM2.md?plain=1#L48)

Done 👍

dennis-tra commented on reports/RFM2.md:
```diff
@@ -0,0 +1,131 @@
+This RFM had short term as well as long term benefits. On the one hand, it answered the aforementioned measurement questions and on the other hand it initiated the development of valuable measurement tooling that can be used for future research. By diving into the data, many more questions arose that led to other RFMs like RFM5 or RFM6. Both of which can be tackeled with the the software that was ...
+
+The measurement results itself reveal a couple of interesting insights. For one, the churn rate in the IPFS network is quite high and does not differ much from previous peer-to-peer networks from over 20 years ago. Translating this finding in an evaluation for network wide parameters like record redundancy, or query concurrency is part of the future work.
+
```
Thanks Yiannis, included! :)

guillaumemichel commented:
> Suggestions welcome for an ioi reference.

I have to say that I don't know what `ioi` is and am now even more curious after reading about `storm`. Is it another virus?

> what do you think? I actually haven't mentioned storm nodes in the report yet.

As it appears on the plots, I think it would make sense to shortly describe what it is.

> I moved the report to results/rfm2-uptime-and-churn.md. I moved the graphs in a respective results/rfm2-uptime-and-churn/ folder. Fine with you @guillaumemichel ?

Perfect for the report!

It would be better to have the graphs in `implementations/rfm2-uptime-and-churn/` or inside a subfolder there. This way we could have a clean `results/` folder containing only the RFMs reports.

> Are you fine with the (arguably long) list of highlights? Call it Highlights? Feel free to remove/add things :)

Looks good! You could even move the Nebula Crawler to the top of the list ;)

>  Where to put the code? See issue(s) above.

Once the plots are in the definitive version, or once the report is in final state, I think that it would be fine to add the [Nebula Crawler](https://github.com/dennis-tra/nebula-crawler) repo as a [submodule](https://devconnected.com/how-to-add-and-update-git-submodules/) inside `implementations/rfm2-uptime-and-churn/`. The submodule would be a snapshot of the state of the Nebula Crawler repo at the time the report is published, and thus even if it doesn't get the last updates, it is still possible to reproduce the results using this snapshot. The submodule could be updated if there is a (major) update in the report. What do you think @dennis-tra @yiannisbot 

Concerning the backup / hosting of Nebula Crawler souce code on PL Github org, we could have an auto-updating fork [Nebula Crawler](https://github.com/dennis-tra/nebula-crawler) hosted on the `protocol` org using this [Github app](https://probot.github.io/apps/pull/).

dennis-tra commented:
> @dennis-tra I just noticed that you removed the graphs from the report. I think it is great to have plots in the report like in the initial version! My comment was about moving the folder where the plots are stored out of the `results/` folder that will be containing RFM reports only (and no images).

Yikes, I moved and renamed the folders but didn't update the references. I did not remove them intentionally. Should be fixed now 👍

guillaumemichel commented on RFMs.md:
```diff
@@ -95,15 +94,14 @@ We have a dashboard that shows the lifetime of documents, differentiated by diff
 * _Prerequisite(s):_ NONE
 * _Value:_ **HIGH**
-* _Report:_ \<insert link to report once work is complete\>
+* _Report:_ [`./reports/RFM2.md`](./reports/RFM2.md)
```
`[./results/rfm2-uptime-and-churn.md](./results/rfm2-uptime-and-churn.md)`
